sdk/identity/azure-identity/src/azure_pipelines_credential.cpp

```diff
@@ -0,0 +1,257 @@
+
+    std::string message = GetCredentialName() + " : "
+        + std::to_string(static_cast<std::underlying_type<decltype(statusCode)>::type>(statusCode))
+        + " response from the OIDC endpoint. Check service connection ID and Pipeline "
+          "configuration. "
+        + response->GetReasonPhrase() + "\n\n" + responseBody;
+    IdentityLog::Write(IdentityLog::Level::Verbose, message);
+
+    throw AuthenticationException(message);
+  }
+
+  json parsedJson;
+  try
+  {
+    parsedJson = Azure::Core::Json::_internal::json::parse(responseBody);
+  }
+  catch (json::exception const&)
+  {
+    std::string message = GetCredentialName() + " : Cannot parse the response string as JSON.";
+    IdentityLog::Write(IdentityLog::Level::Verbose, message);
+
+    throw AuthenticationException(message);
+  }
+
+  const std::string oidcTokenPropertyName = "oidcToken";
+  if (!parsedJson.contains(oidcTokenPropertyName) || !parsedJson[oidcTokenPropertyName].is_string())
+  {
+    std::string message = GetCredentialName()
+        + " : OIDC token not found in response. \nSee Azure::Core::Diagnostics::Logger for details "
+          "(https://aka.ms/azsdk/cpp/identity/troubleshooting).";
+    IdentityLog::Write(IdentityLog::Level::Verbose, message);
+    throw AuthenticationException(message);
+  }
+  return parsedJson[oidcTokenPropertyName].get<std::string>();
+}
+
+AzurePipelinesCredential::~AzurePipelinesCredential() = default;
+
+AccessToken AzurePipelinesCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  if (!m_tokenCredentialImpl)
+  {
+    auto const AuthUnavailable = GetCredentialName() + " authentication unavailable. ";
+
+    IdentityLog::Write(
+        IdentityLog::Level::Warning,
+        AuthUnavailable + "See earlier " + GetCredentialName() + " log messages for details.");
+
+    throw AuthenticationException(
+        AuthUnavailable + "Azure Pipelines environment is not set up correctly.");
+  }
+
+  auto const tenantId = TenantIdResolver::Resolve(
+      m_clientCredentialCore.GetTenantId(),
+      tokenRequestContext,
+      m_clientCredentialCore.GetAdditionallyAllowedTenants());
+
+  auto const scopesStr
+      = m_clientCredentialCore.GetScopesString(tenantId, tokenRequestContext.Scopes);
+
+  Azure::Core::Http::Request oidcRequest = CreateOidcRequestMessage();
+  std::unique_ptr<RawResponse> response = m_httpPipeline.Send(oidcRequest, context);
+
+  if (!response)
+  {
+    throw AuthenticationException(
+        GetCredentialName() + " couldn't send OIDC token request: null response.");
+  }
+
+  auto const bodyStream = response->ExtractBodyStream();
+  auto const bodyVec = bodyStream ? bodyStream->ReadToEnd(context) : response->GetBody();
+  auto const responseBody
+      = std::string(reinterpret_cast<char const*>(bodyVec.data()), bodyVec.size());
+
+  std::string assertion = GetOidcTokenResponse(response, responseBody);
```

Comment:
Move these lines to be after the line 238, so that if the token cache has the valid on-expired token token, we won't end up doing all this - sending a request and parsing, only to discard later.

Reply:
There are **two** requests made in this credential:
1) Request sent to Oidc, which fetches the assertion. (not cached)
2) Request sent to microsoft, which uses the assertion, and fetches the access token. (cached)

The first request, and its "token" isn't cached, and hence it isn't part of the token caching. The second request, which actually fetches the `AccessToken`, is being cached, like all the other credentials.

Reply:
It is not about caching the OIDC.

The first time it should work just as fine - it will try to hit the cache for the given `scopesStr` and `tenantId`, the cache will be empty. Then it will call the lambda which will send the request to Oidc, fetch the assertion, will put that assertion to the request to Microsoft, will fetch the token, put it in cache, and return it.

The second time, it will see that it has non-expired token for the given `scopesStr` and `tenantId`, and will return it.
Only that currently, before even looking into cache, you will still send a request to Oidc, unconditionally, parse the response, get the assertion, and then return the cached token value without ever using the assertion you just got. I.e. you are sending requests to Oidc every time during `GetToken()`, and you are not using that value most of the time (when the cached token is good).